test/suite/test_prepare_hs03.py

```diff
@@ -0,0 +1,183 @@
+    # Force a small cache.
+    conn_config = 'cache_size=50MB'
+
+    # Create a small table.
+    uri = "table:test_prepare_hs03"
+
+    def corrupt_file(self):
+        filename="test_prepare_hs03.wt"
+        self.assertEquals(os.path.exists(filename), True)
+
+        with open(filename, 'r+') as log:
+            log.seek(1024)
+            log.write('Bad!' * 1024)
+
+    def corrupt_salvage_verify(self):
+        self.corrupt_file()
+        self.session.salvage(self.uri, "force")
+        self.session.verify(self.uri, None)
+
+    def prepare_updates(self, ds, nrows, nsessions, nkeys):
+        # Insert some records with commit timestamp, corrupt file and call salvage, verify before checkpoint.
+
+        # Commit some updates to get eviction and history store fired up
+        bigvalue1 = b"bbbbb" * 2
+        cursor = self.session.open_cursor(self.uri)
+        for i in range(1, nkeys):
+            self.session.begin_transaction('isolation=snapshot')
+            tkey = 'C' + ds.key(nrows + i)
+            cursor.set_key(tkey)
+            cursor.set_value(bigvalue1)
+            self.assertEquals(cursor.insert(), 0)
+            self.session.commit_transaction('commit_timestamp=' + timestamp_str(1))
+        cursor.close()
+
+        # Corrupt the table, Call salvage to recover data from the corrupted table and call verify
+        self.corrupt_salvage_verify()
+
+        # Call checkpoint
+        self.session.checkpoint()
+
+        # Corrupt the table, Call salvage to recover data from the corrupted table and call verify
+        self.corrupt_salvage_verify()
+
+        # Have prepared updates in multiple sessions. This should ensure writing
+        # prepared updates to the history store
+        sessions = [0] * nsessions
+        cursors = [0] * nsessions
+        bigvalue2 = b"ccccc" * 2
+        for j in range (0, nsessions):
+            sessions[j] = self.conn.open_session()
+            sessions[j].begin_transaction('isolation=snapshot')
+            cursors[j] = sessions[j].open_cursor(self.uri)
+            # Each session will update many consecutive keys.
+            start = (j * nkeys)
+            end = start + nkeys
+            for i in range(start, end):
+                tkey = 'P' + ds.key(nrows + i)
+                cursors[j].set_key(tkey)
+                cursors[j].set_value(bigvalue2)
+                self.assertEquals(cursors[j].insert(), 0)
+            sessions[j].prepare_transaction('prepare_timestamp=' + timestamp_str(4))
+
+        # Test if we can read prepared updates from the history store.
+        cursor = self.session.open_cursor(self.uri)
+        self.session.begin_transaction('read_timestamp=' + timestamp_str(2))
+        for i in range(1, nsessions * nkeys):
+            tkey = 'P' + ds.key(nrows + i)
+            cursor.set_key(tkey)
+            # The search should fail i.e, WT_NOTFOUND
+            self.assertEqual(cursor.search(), wiredtiger.WT_NOTFOUND)
```

Comment:
If you end up deleting the prefixes for the keys, the search here will succeed, and the correctness test will need to be whether the committed value is visible, rather than the prepared value.  If you go this route, you won't need to do a final scan searching for keys inserted with the commit (the loop at line 129), since you they will be all the same keys.  Although the scan after the simulated crash would be still be good.

Reply:
Updated the code and added some correctness tests to check that committed value is visible and prepared value not visible.